src/NuGet.Core/NuGet.LibraryModel/DownloadDependency.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright(c) .NET Foundation.All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using NuGet.Shared;
+using NuGet.Versioning;
+
+namespace NuGet.LibraryModel
+{
+    public class DownloadDependency : IEquatable<DownloadDependency>, IComparable<DownloadDependency>
```

Comment:
I didn't see where in the code it was being used. There were no explicit calls to CompareTo or calls to something like List.Sort or Array.Sort, but if this class is being added to a sorted collection, then yes, it's used.

The reason I believe it's not technically correct is because semantically CompareTo is about sorting, whereas hash codes are only about equality (with the caveat of collisions). If the implementation of the hash code happens to produce a sortable value, that's something that everyone reading the code must understand but is not fundamentally inherit in the concept of a hash code.

Furthermore, looking at the implementation of our hash code generation:
https://github.com/NuGet/NuGet.Client/blob/21f2b07f2c2e84afd6602c6743d356dae8880c0c/src/NuGet.Core/NuGet.Commands/Internal/HashCodeCombiner.cs#L52-L55

The `<< 5` part of this makes me think that there will be issues when any part of the version has a value larger than or equal to (1 << 6), or 64. Even if testing shows that 1.88888.0 is less than 2.0.0 with this code, I feel like it's most likely by accident rather than by design.

I agree it's out of scope of this PR, but once you see that VersionRange and Version both use HashCodeCombiner's results to sort, it's not at all intuitive that this will actually produce correct results.

Reply:
> If the implementation of the hash code happens to produce a sortable value, that's something that everyone reading the code must understand but is not fundamentally inherit in the concept of a hash code.

By the definition of hash code, it will always generate something sortable. 

> Even if testing shows that 1.88888.0 is less than 2.0.0 with this code, I feel like it's most likely by accident rather than by design.

The sorted order does not have to make sense to the naked eye, as long it's consistent. 

I have created an issue to address this https://github.com/NuGet/Home/issues/7728.